### PR TITLE
FN_DEADLINE for both http and json

### PR DESCRIPTION
**Pull request by denismakogon**

### CODE REVIEW

File: fdk.go [hunk 1]
```diff
@@ -175,12 +176,16 @@ func doJSONOnce(handler Handler, ctx context.Context, in io.Reader, out io.Write
 	json.NewEncoder(out).Encode(jsonResponse)
 }
 
-func doHTTPOnce(handler Handler, ctx context.Context, in io.Reader, out io.Writer, buf *bytes.Buffer, hdr http.Header) {
-	// TODO we need to set deadline on ctx here (need FN_DEADLINE header)
-	// for now, just get a new ctx each go round
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+func CtxWithDeadline(parent context.Context, fnDeadline string) (context.Context, context.CancelFunc) {
```
Comment (rdallman) [inline fdk.go]: can we lower case this? don't really want it to pollute public API in docs, we're giving them a handle on a context already, they shouldn't need this.

File: fdk.go [hunk 2]
```diff
@@ -175,12 +176,16 @@ func doJSONOnce(handler Handler, ctx context.Context, in io.Reader, out io.Write
...
+	ctx, cancel := context.WithCancel(parent)
+	t, err := time.Parse(time.RFC3339, fnDeadline)
+	if err == nil {
+		ctx, cancel = context.WithDeadline(ctx, t)
+	}
+	return ctx, cancel
```
Comment (rdallman) [inline fdk.go]: slightly simpler control flow to do: 

```go
if err == nil {
  return context.WithDeadline(ctx, t)
}
return context.WithCancel(ctx)
```

it's also unconventional to name the context `parent` in the signature, please change (go lint would yell)

File: fdk.go [hunk 3]
```diff
@@ -176,11 +176,11 @@ func doJSONOnce(handler Handler, ctx context.Context, in io.Reader, out io.Write
 	json.NewEncoder(out).Encode(jsonResponse)
 }
 
-func CtxWithDeadline(parent context.Context, fnDeadline string) (context.Context, context.CancelFunc) {
-	ctx, cancel := context.WithCancel(parent)
+func ctxWithDeadline(ctx context.Context, fnDeadline string) (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(ctx)
```
Comment (rdallman) [inline fdk.go]: go vet would pick up that this cancel function is leaked in one of the control flow paths fwiw (only the child context gets canceled if we find a deadline, not the cancel's). see prev suggestion

Comment (denismakogon) [reply fdk.go]: fixed.

**Comment by rdallman** [comment]
sweet, thanks. cc @skinowski

**Pull request merged**